ctapipe/tools/tests/test_merge.py

```diff
@@ -122,3 +124,24 @@ def test_allowed_tels(tmp_path, dl1_file, dl1_proton_file):
 
     s = SubarrayDescription.from_hdf(output)
     assert s.tel.keys() == {1, 2}
+
+
+def test_split_datasets_by(tmp_path, dl1_file, dl1_proton_file):
+    from ctapipe.tools.dl1_merge import MergeTool
+
+    # create file 'split_datasets_by' option
+    output = tmp_path / "split_datasets_by_tel_type.dl1.h5"
+
+    ret = run_tool(
+        MergeTool(),
+        argv=[
+            str(dl1_file),
+            str(dl1_proton_file),
+            f"--output={output}",
+            "--split_datasets_by=tel_type",
+            "--overwrite",
+        ],
+        cwd=tmp_path,
+    )
+    assert ret == 0
+
```

Comment:
I'd suggest adding to this test a check that indeed the output has the right datasets in it (e.g. that if split by tel type, then open the file and check that there is a dataset named by the tel type).

Basically, the test should ensure that the functionality is there.  For example, if later simply ignore the split by by tel_type option and always split only by tel_id,  this test should fail, but right now it does not.

Reply:
Should be enough to do something like 
```py3
with tables.open_file(output) as h5file:
    assert h5file.get_node("/dl1/event/telescope/parameters/LST_LST_LSTCam") is not None
```

Reply:
> ```
> def test_allowed_tels_and_split_datasets_by(tmp_path, gamma_dl1_path):
>     from ctapipe.tools.dl1_merge import MergeTool
>     from ctapipe.io import read_table
> 
>     # create a second file so we can test the split by tel type
>     output = tmp_path / "split_datasets_by_tel_type.dl1.h5"
> 
>     ret = run_tool(
>         MergeTool(),
>         argv=[
>             str(gamma_dl1_path),
>             f"--output={output}",
>             "--allowed-tels=[1,2,3,4]",
>             "--split_datasets_by=tel_type",
>             "--overwrite",
>         ],
>         cwd=tmp_path,
>     )
>     assert ret == 0
> 
>     original_table_length = 0
>     for tel_id in range(1,5):
>         original_table_length += len(read_table(str(gamma_dl1_path), f"/dl1/event/telescope/parameters/tel_{tel_id:03d}"))
> 
>     merged_table_length = len(read_table(output, f"/dl1/event/telescope/parameters/LST_LST_LSTCam"))
> 
>     assert original_table_length == merged_table_length
> ```
> 
> However, this test is failing (works just fine outside of pytest with handmade dl1 files):
> 
> ```
> >           original_table_length += len(read_table(str(gamma_dl1_path), f"/dl1/event/telescope/parameters/tel_{tel_id:03d}"))
> 
> ctapipe/tools/tests/test_merge.py:130: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> ctapipe/io/astropy_helpers.py:59: in read_table
>     table = h5file.get_node(path)
> ../../anaconda3/envs/cta/lib/python3.7/site-packages/tables/file.py:1642: in get_node
>     node = self._get_node(nodepath)
> ../../anaconda3/envs/cta/lib/python3.7/site-packages/tables/file.py:1590: in _get_node
>     node = self._node_manager.get_node(nodepath)
> ../../anaconda3/envs/cta/lib/python3.7/site-packages/tables/file.py:432: in get_node
>     node = self.node_factory(key)
> ../../anaconda3/envs/cta/lib/python3.7/site-packages/tables/group.py:1178: in _g_load_child
>     node_type = self._g_check_has_child(childname)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> self = <closed tables.group.RootGroup at 0x7f5d2fe21cb0>, name = '/dl1/event/telescope/parameters/tel_001'
> 
>     def _g_check_has_child(self, name):
>         """Check whether 'name' is a children of 'self' and return its type."""
>     
>         # Get the HDF5 name matching the PyTables name.
>         node_type = self._g_get_objinfo(name)
>         if node_type == "NoSuchNode":
>             raise NoSuchNodeError(
>                 "group ``%s`` does not have a child named ``%s``"
> >               % (self._v_pathname, name))
> E           tables.exceptions.NoSuchNodeError: group ``/`` does not have a child named ``/dl1/event/telescope/parameters/tel_001``
> 
> ../../anaconda3/envs/cta/lib/python3.7/site-packages/tables/group.py:395: NoSuchNodeError
> ```

I think the problem is that the node is not created when no triggered event survived? Neither tel_001 nor LST_LST_LSTCam can be found in the "split by tel_ids" files and  "split by tel_types" files, respectively, for the test data.

Reply:
There are events in there, so at least one telescope type is there. You can check which ones are there by inspecting the created file with vitables, something like `vitables /tmp/pytest-of-$USER/pytest-current/<test_name>current/<file>`

Reply:
Thanks @maxnoe! Triggered events were found for tel id 25, 125 (FlashCam), and 130 (NectarCam). I updated the merger test accordingly and it passes locally.